trunk/vzic/vzic.1
." Text automatically generated by txt2man-1.4.7
.TH VZIC 1 "October 27, 2013" "vzic-1.3" "Linux Reference Manual"
.SH NAME
\fBvzic \fP- convert Olson timezone database files to iCalendar format
\fB
.SH SYNOPSIS
.nf
.fam C
\fBvzic\fP [\fB--dump\fP] [\fB--dump-changes\fP] [\fB--no-rrules\fP] [\fB--no-rdates\fP] [\fB--pure\fP]
[\fB--output-dir\fP <directory>] [\fB--url-prefix\fP <url>] [\fB--olson-dir\fP <directory>]
.fam T
.fi
.SH DESCRIPTION
\fBvzic\fP is a program to convert the Olson timezone database files into
VTIMEZONE files compatible with the iCalendar specification (RFC2445).
.SH OPTIONS
.TP
.B
\fB--pure\fP
Output the perfect VCALENDAR data, which Outlook won't parse
as it has problems with certain iCalendar constructs.
.TP
.B
\fB--output-dir\fP
specify where to output all the files beneath. The
default is the current directory. 
.TP
.B
\fB--url-prefix\fP
Used as the base for the TZURL property in each
VTIMEZONE. The default is to not output TZURL properties. 
.PP
Debugging Options: 
.TP
.B
\fB--dump\fP
Dump the Rule and Zone data that we parsed from the Olson
timezone files. This is used to test the parsing code. 
.TP
.B
\fB--dump-changes\fP
Dumps a list of times when each timezone changed,
and the new local time offset from UTC. 
.TP
.B
\fB--no-rrules\fP
Don't output RRULE properties in the VTIMEZONEs. Instead
it will just output RDATEs for each year up to a certain year. 
.TP
.B
\fB--no-rdates\fP
Don't output multiple RDATEs in a single VTIMEZONE
component. Instead they will be output separately. 
.SH COPYRIGHT

Copyright (C) 2000-2001 Ximian, Inc.
Copyright (C) 2003 Damon Chaplin.
.PP
Author: Damon Chaplin <damon@gnome.org>
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.


changelog
ietfdb (6.55.1) ietf; urgency=medium

  This is a bugfix release, with some smaller enhancements.  From the commit
  log:

  * Don't deliver spurious yang extraction errors for drafts where no yang 
    modules were found.

  * Added circular parent relationship detection and logging in the
    Group.is_decendant_of() method, and form validation error on loop in the
    group edit form data cleaning.

  * Increased the test crawlers verbose output.

  * Added a guard around rendering of group.ad_role, which isn't always set.

  * Refined the appearance and wording of the 'Manage document's adoption' 
    button, to make it possible for chairs to correct an adoption in the
    wrong group.

  * Replaced the use of unaccent.asciify(), which has similar functionality 
    as unidecode.unidecode().  Changed the draft parser to work exclusively 
    with unicode text, which both makes the removal of unaccent easier, and 
    takes us closer to Py35 compatibility.  Adjusted callers of the draft 
    parser to send in unicode.

  * Removed code which has been marked as unreachable since February/March 
    2017 and has not triggering any usage indications.

  * Changed the display of people with roles on the various group pages to 
    link to both their profile page and their email address.

  * Improved the handling of surname particles, (prompted by failure to 
    handle the name of Niels ten Oever correctly).

  * Made changes to the can_adopt_draft() logic, to make it possible for a 
    chair of multiple groups to correct the mistake if adoption was made for 
    the wrong group.

  * Merged in [13615] and [13627]from rcross@amsl.com:
    Changed views that use GET requests to delete objects to use POST.
    Fixes #1796.

  * Make sure wordwrap() and friend works as intended if they are used as 
    template filters and given string arguments.

 -- Henrik Levkowetz <henrik@levkowetz.com>  20 Jun 2017 13:43:31 +0000


ietfdb (6.55.0) ietf; urgency=medium

  **Improved Yang validation support**

  This release updates and expands the Yang support in the datatracker.  In
  addition to checking yang modules in submitted drafts using pyang, it now
  also does a yanglint check.  The yang validation results are available by
  clicking on the red or green yin/yang symbols which show for drafts that
  contain yang models.

  For drafts where one or more models were found at submission time, the pyang
  and yanglint checks are re-run once per day, to catch newly arrived models
  in the library directories (both rfc and draft models).

  Additional details, from the commit log:

  * Added a new yang checker, 'yanglint', to the existing Yang checker class,
    in addition to the existing 'pyang' checker.

  * Added modal overlay displays showing the yang check results every place
    the yin/yang symbol is shown (red or green) to indicate the presencee and
    result of yang checks.  Added a Yang Validation: line in the document
    meta-information section on the document's page in the datatracker.

  * Added the result of the xym extaction to the yang check results, to make
    extration failures visible.

  * Added the version of the used xym, pyang, and yanglint commands to the
    check results.

  * Added an action to move successfully extracted and validated modules to
    the module library directories immediately on submission.

  * Added the xym repository as an svn:external component, rather than listing
    it in requirements.txt, as there has been delays of many months between
    essential features appearing in the repository, and an actual release of
    same.  We may get occasional buildbot failures if broken code is pulled in
    from the repository, but better that than the functionality failure of
    severely outdated componets.

  * Added a new management command to re-run yang validation for active drafts
    for which yang modules were found at submission time, in order to pick up
    imported models which may have arrived in the model libraries after the
    draft's submission.  Run daily from bin/daily.

  * Added a table to hold version information for external commands.  The yang
    checker output should include the version information of the used
    checkers, but seems unnecessary to run each command with its --version
    switch every time we check a module...

  * Added a new management command to collect version information for external
    commands on demand.  To be run daily from bin/daily.

  * Added tests to verify that xym, pyang and yanglint information is
    available on the submission confirmation page, and updated the yang module
    contained in the test document to validate under both pyang and yanglint.

  * Updated admin.py and resource.py files as needed.

 -- Henrik Levkowetz <henrik@levkowetz.com>  16 Jun 2017 09:38:25 +0000


ietfdb (6.54.1) ietf; urgency=medium

  **Bugfixes, floor indications on the agenda**

  This is a bugfix release, with a number of smaller fixes.  It also
  introduces floor indications on the IETF meeting agenda page.  From the
  commit log:

  * Added a guard against trying to get the room ID from a timeslot without 
    location in the secretariat schedule view.

  * Merged in [13588] from rcross@amsl.com:
    Initialize Empty-Schedule with special sessions from previous meeting on
    creation.  Fixes issue #1500.

  * Merged in [13586] from rjsparks@nostrum.com:
    Workaround for an issue with the schedule editor by ensuring no sessions
    remain scheduled in a room whose purpose has changed to something other
    than holding sessions (such as becoming unavailable).  

  * A first go at adding floor labels to the agenda pages.  The placement 
    of the labels are complicated by available whitespace beeing in different 
    horizontal positions for different types of lines.  Adding another column 
    will push the session title out of the screen on small devices.


trunk/vzic/vzic.c
/*
 * Vzic - a program to convert Olson timezone database files into VZTIMEZONE
 * files compatible with the iCalendar specification (RFC2445).
 *
 * Copyright (C) 2000-2001 Ximian, Inc.
 * Copyright (C) 2003 Damon Chaplin.
 *
 * Author: Damon Chaplin <damon@gnome.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "vzic.h"
#include "vzic-parse.h"
#include "vzic-dump.h"
#include "vzic-output.h"


/*
 * Global command-line options.
 */

/* By default we output Outlook-compatible output. If --pure is used we
   output pure output, with no changes to be compatible with Outlook. */
gboolean VzicPureOutput			= FALSE;

gboolean VzicDumpOutput			= FALSE;
gboolean VzicDumpChanges		= FALSE;
gboolean VzicDumpZoneNamesAndCoords	= TRUE;
gboolean VzicDumpZoneTranslatableStrings= TRUE;
gboolean VzicNoRRules			= FALSE;
gboolean VzicNoRDates			= FALSE;
char*    VzicOutputDir			= "zoneinfo";
char*    VzicUrlPrefix                  = NULL;
char*    VzicOlsonDir                   = OLSON_DIR;

GList*	 VzicTimeZoneNames		= NULL;

static void	convert_olson_file		(char		*olson_file);

static void	usage				(void);

static void	free_zone_data			(GArray		*zone_data);
static void	free_rule_array			(gpointer	 key,
						 gpointer	 value,
						 gpointer	 data);
static void	free_link_data			(gpointer	 key,
						 gpointer	 value,
						 gpointer	 data);


int
main				(int		 argc,
				 char		*argv[])
{
  int i;
  char directory[PATHNAME_BUFFER_SIZE];
  char filename[PATHNAME_BUFFER_SIZE];
  GHashTable *zones_hash;

  /*
   * Command-Line Option Parsing.
   */
  for (i = 1; i < argc; i++) {
    /*
     * User Options.
     */

    /* --pure: Output the perfect VCALENDAR data, which Outlook won't parse
       as it has problems with certain iCalendar constructs. */
    if (!strcmp (argv[i], "--pure"))
      VzicPureOutput = TRUE;

    /* --output-dir: specify where to output all the files beneath. The
       default is the current directory. */
    else if (argc > i + 1 && !strcmp (argv[i], "--output-dir"))
      VzicOutputDir = argv[++i];

    /* --url-prefix: Used as the base for the TZURL property in each
       VTIMEZONE. The default is to not output TZURL properties. */
    else if (argc > i + 1 && !strcmp (argv[i], "--url-prefix")) {
      int length;
      VzicUrlPrefix = argv[++i];
      /* remove the trailing '/' if there is one */
      length = strlen (VzicUrlPrefix);
      if (VzicUrlPrefix[length - 1] == '/')
          VzicUrlPrefix[length - 1] = '\0';
    }

    else if (argc > i + 1 && !strcmp (argv[i], "--olson-dir")) {
      VzicOlsonDir = argv[++i];
    }

    /*
     * Debugging Options.
     */

    /* --dump: Dump the Rule and Zone data that we parsed from the Olson
       timezone files. This is used to test the parsing code. */
    else if (!strcmp (argv[i], "--dump"))
      VzicDumpOutput = TRUE;

    /* --dump-changes: Dumps a list of times when each timezone changed,
       and the new local time offset from UTC. */
    else if (!strcmp (argv[i], "--dump-changes"))
      VzicDumpChanges = TRUE;

    /* --no-rrules: Don't output RRULE properties in the VTIMEZONEs. Instead
       it will just output RDATEs for each year up to a certain year. */
    else if (!strcmp (argv[i], "--no-rrules"))
      VzicNoRRules = TRUE;

    /* --no-rdates: Don't output multiple RDATEs in a single VTIMEZONE
       component. Instead they will be output separately. */
    else if (!strcmp (argv[i], "--no-rdates"))
      VzicNoRDates = TRUE;

    else
      usage ();
  }

  /*
   * Create any necessary directories.
   */
  ensure_directory_exists (VzicOutputDir);

  if (VzicDumpOutput) {
    /* Create the directories for the dump output, if they don't exist. */
    sprintf (directory, "%s/ZonesVzic", VzicOutputDir);
    ensure_directory_exists (directory);
    sprintf (directory, "%s/RulesVzic", VzicOutputDir);
    ensure_directory_exists (directory);
  }

  if (VzicDumpChanges) {
    /* Create the directory for the changes output, if it doesn't exist. */
    sprintf (directory, "%s/ChangesVzic", VzicOutputDir);
    ensure_directory_exists (directory);
  }

  /*
   * Convert the Olson timezone files.
   */
  convert_olson_file ("africa");
  convert_olson_file ("antarctica");
  convert_olson_file ("asia");
  convert_olson_file ("australasia");
  convert_olson_file ("europe");
  convert_olson_file ("northamerica");
  convert_olson_file ("southamerica");

  /* These are backwards-compatability and weird stuff. */
#if 0
  convert_olson_file ("backward");
  convert_olson_file ("etcetera");
  convert_olson_file ("leapseconds");
  convert_olson_file ("pacificnew");
  convert_olson_file ("solar87");
  convert_olson_file ("solar88");
  convert_olson_file ("solar89");
#endif

  /* This doesn't really do anything and it messes up vzic-dump.pl so we
     don't bother. */
#if 0
  convert_olson_file ("factory");
#endif

  /* This is old System V stuff, which we don't currently support since it
     uses 'min' as a Rule FROM value which messes up our algorithm, making
     it too slow and use too much memory. */
#if 0
  convert_olson_file ("systemv");
#endif

  /* Output the timezone names and coordinates in a zone.tab file, and
     the translatable strings to feed to gettext. */
  if (VzicDumpZoneNamesAndCoords) {
    sprintf (filename, "%s/zone.tab", VzicOlsonDir);
    zones_hash = parse_zone_tab (filename);

    dump_time_zone_names (VzicTimeZoneNames, VzicOutputDir, zones_hash);
  }

  return 0;
}


static void
convert_olson_file		(char		*olson_file)
{
  char input_filename[PATHNAME_BUFFER_SIZE];
  GArray *zone_data;
  GHashTable *rule_data, *link_data;
  char dump_filename[PATHNAME_BUFFER_SIZE];
  ZoneData *zone;
  int i, max_until_year;

  sprintf (input_filename, "%s/%s", VzicOlsonDir, olson_file);

  parse_olson_file (input_filename, &zone_data, &rule_data, &link_data,
		    &max_until_year);

  if (VzicDumpOutput) {
    sprintf (dump_filename, "%s/ZonesVzic/%s", VzicOutputDir, olson_file);
    dump_zone_data (zone_data, dump_filename);

    sprintf (dump_filename, "%s/RulesVzic/%s", VzicOutputDir, olson_file);
    dump_rule_data (rule_data, dump_filename);
  }

  output_vtimezone_files (VzicOutputDir, zone_data, rule_data, link_data,
			  max_until_year);

  free_zone_data (zone_data);
  g_hash_table_foreach (rule_data, free_rule_array, NULL);
  g_hash_table_destroy (rule_data);
  g_hash_table_foreach (link_data, free_link_data, NULL);
  g_hash_table_destroy (link_data);
}


static void
usage				(void)
{
  fprintf (stderr, "Usage: vzic [--dump] [--dump-changes] [--no-rrules] [--no-rdates] [--pure] [--output-dir <directory>] [--url-prefix <url>] [--olson-dir <directory>]\n");

  exit (1);
}




/*
 * Functions to free the data structures.
 */

static void
free_zone_data			(GArray		*zone_data)
{
  ZoneData *zone;
  ZoneLineData *zone_line;
  int i, j;

  for (i = 0; i < zone_data->len; i++) {
    zone = &g_array_index (zone_data, ZoneData, i);

    g_free (zone->zone_name);

    for (j = 0; j < zone->zone_line_data->len; j++) {
      zone_line = &g_array_index (zone->zone_line_data, ZoneLineData, j);

      g_free (zone_line->rules);
      g_free (zone_line->format);
    }

    g_array_free (zone->zone_line_data, TRUE);
  }

  g_array_free (zone_data, TRUE);
}


static void
free_rule_array			(gpointer	 key,
				 gpointer	 value,
				 gpointer	 data)
{
  char *name = key;
  GArray *rule_array = value;
  RuleData *rule;
  int i;

  for (i = 0; i < rule_array->len; i++) {
    rule = &g_array_index (rule_array, RuleData, i);

    if (!rule->is_shallow_copy) {
      g_free (rule->type);
      g_free (rule->letter_s);
    }
  }

  g_array_free (rule_array, TRUE);

  g_free (name);
}


static void
free_link_data			(gpointer	 key,
				 gpointer	 value,
				 gpointer	 data)
{
  GList *link = data;

  g_free (key);

  while (link) {
    g_free (link->data);
    link = link->next;
  }

  g_list_free (data);
}

